packages/wdio-sauce-service/src/launcher.ts

```diff
@@ -0,0 +1,121 @@
+     */
+    async onPrepare (config: SauceLabsOptions, capabilities: DesiredCapabilities | Array<DesiredCapabilities>) {
+        if (!this._options.sauceConnect) {
+            return
+        }
+
+        const sauceConnectOpts = this._options.sauceConnectOpts || {}
+        const sauceConnectTunnelIdentifier = (
+            sauceConnectOpts.tunnelIdentifier ||
+            /**
+             * generate random identifier if not provided
+             */
+            `SC-tunnel-${Math.random().toString().slice(2)}`)
+
+        sauceConnectOpts.noAutodetect = true
+        sauceConnectOpts.tunnelIdentifier = sauceConnectTunnelIdentifier
+
+        let endpointConfigurations = {}
+        if (this._options.scRelay) {
+            log.warn(SC_RELAY_DEPCRECATION_WARNING)
+
+            const scRelayPort = sauceConnectOpts.sePort || 4445
+            sauceConnectOpts.sePort = scRelayPort
+            endpointConfigurations = {
+                protocol: 'http',
+                hostname: 'localhost',
+                port: scRelayPort
+            }
+        }
+
+        const prepareCapability = makeCapabilityFactory(sauceConnectTunnelIdentifier, endpointConfigurations)
+
+        if (Array.isArray(capabilities)) {
+            for (const capability of capabilities) {
+                prepareCapability(capability)
+            }
+        } else {
+            for (const browserName in capabilities) {
+                const browserCap = capabilities[browserName as keyof typeof capabilities] as Record<string, any>
+                prepareCapability(browserCap['capabilities'])
+            }
+        }
+
+        /**
+         * measure SC boot time
+         */
+        const obs = new PerformanceObserver((list) => {
+            const entry = list.getEntries()[0]
+            log.info(`Sauce Connect successfully started after ${entry.duration}ms`)
+        })
+        obs.observe({ entryTypes: ['measure'], buffered: false })
+
+        performance.mark('sauceConnectStart')
+        this._sauceConnectProcess = await this._api.startSauceConnect(sauceConnectOpts)
+        performance.mark('sauceConnectEnd')
+        performance.measure('bootTime', 'sauceConnectStart', 'sauceConnectEnd')
+    }
+
+    /**
+     * shut down sauce connect
+     */
+    onComplete () {
+        if (!this._sauceConnectProcess) {
+            return
+        }
+
+        return this._sauceConnectProcess.close()
+    }
+
+    get options(): SauceServiceConfig {
+        return this._options
+    }
+
+    get api(): SauceLabs {
+        return this._api
+    }
+
+    get sauceConnectProcess(): SauceConnectInstance | undefined {
+        return this._sauceConnectProcess
+    }
+
+    set options(value: SauceServiceConfig) {
+        this._options = value
+    }
+
+    set api(value: SauceLabs) {
+        this._api = value
+    }
+
+    set sauceConnectProcess(value: SauceConnectInstance | undefined) {
+        this._sauceConnectProcess = value
+    }
```

Comment:
What is the reason for using getters/setters instead of public variables?

Reply:
Good question, I think that's highly debatable, in languages like Java and C# they ate considered to be best practice, which is why they were added to JS in the first place. Used like this there isn't really a difference, but it could allow for example to return something in a different format, to return a copy of an object to keep encapsulation etc. For the sake of converting it to TS I just added them as is to not have breaking changes, but it could allow further customization for future properties.

Reply:
Got it. Given that we don't do this across the code base, let's remove the getters/setters and use public properties instead.

Reply:
Alright!